File: C++/Step19-Binary Search/2805.cpp
//
//  2805.cpp
//  Step19-Binary Search
//
//  Created by 김예빈 on 2019. 9. 29..
//  Copyright © 2019년 김예빈. All rights reserved.
//

#include <iostream>
using namespace std;

int main(int argc, const char * argv[]) {
    int n, m;
    long long result = 0, mid, left = 0, right = 0, sum;
    
    cin >> n >> m;
    
    int arr[n];
    
    for(int i = 0; i < n; i++) {
        cin >> arr[i];
        
        if(right < arr[i])
            right = arr[i];
    }
    
    while(left <= right) {
        mid = (left + right) / 2;
        sum = 0;
        
        for(int i = 0; i < n; i++)
            if(mid < arr[i])
                sum += arr[i] - mid;
        
        if(sum >= m){
            if(result < mid)
                result = mid;
            
            left = mid + 1;
        } else {
            right = mid - 1;
        }
    }
    
    cout << result;
    
    return 0;
}

File: BaekJoon Online Judge/Swift/Step26-Binary Search/2805.swift
//
//  2805.swift
//  Step20-Binary Search
//
//  Created by Yebin Kim on 2020/01/06.
//  Copyright © 2020 김예빈. All rights reserved.
//

import Foundation

var input = readLine()!.split(separator: " ").map{ Int($0)! }
var N = input[0]
var M = input[1]

var trees = readLine()!.split(separator: " ").map{ Int($0)! }.sorted()
var answer = 0

var left = 0
var right = trees.last!

while left <= right {
    let mid = (left + right) / 2
    var len = 0
    
    for tree in trees where tree > mid {
        len += tree - mid
    }
    
    if len < M {
        right = mid - 1
    } else if len >= M {
        left = mid + 1
        answer = mid
    } else {
        break
    }
}

print(answer)
